docs/design/resource-states.md

```diff
@@ -0,0 +1,222 @@
+* When a failure occurs, it should be clear to the user whether that failure is transient and the operator can self-recover.
+
+### Examining other projects like ASO
+A quick look across the field of projects similar to ASO suggests that there is a relatively standard approach to solving this problem:
+
+* Crossplane reports status through a [Ready condition](https://crossplane.io/docs/v1.3/reference/troubleshoot.html#resource-status-and-conditions)
+* ACK reports status through a [variety of conditions](https://github.com/aws-controllers-k8s/runtime/blob/8191f606c5975d5ba0a1433fab695533980733ba/apis/core/v1alpha1/conditions.go), including `ACK.Adopted`, `ACK.resourceSynced`, `ACK.Terminal`, etc.
+* Cluster API originally used a `phase` and `failureReason`/`failureMessage` pattern, but has since [moved to use conditions](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200506-conditions.md) and is deprecating the old pattern.
+* Pod uses a combination of [phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) and [conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions), including `PodScheduled`, `ContainersReady`, `Initalized`, and `Ready`.
+
+It seems that most of these (and many core Kubernetes resources) use a `Ready` condition to specify the status.
+
+### More on conditions
+These best practices were gathered from [conditions API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) as of July 2021.
+
+> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See #7856 for more details and discussion.
+
+> In condition types, and everywhere else they appear in the API, Reason is intended to be a one-word, CamelCase representation of the category of cause of the current status, and Message is intended to be a human-readable phrase or sentence, which may contain specific details of the individual occurrence. Reason is intended to be used in concise output, such as one-line kubectl get output, and in summarizing occurrences of causes, whereas Message is intended to be presented to users in detailed status explanations, such as kubectl describe output.
```

Comment:
Do you mean `camelCase` (leading lowercase with uppercase word separators) or `PascalCase` (leading uppercase with uppercase word separators)?

Also suggest "... Reason is intended to be a single compound-word, ... "

Reply:
Though, if this is a quote ... keep it intact ...

Reply:
This is a quote, and camelCase is apparently a Microsoft-ism according to [Wikipedia](https://en.wikipedia.org/wiki/Camel_case):
> Camel case is often used as a naming convention in computer programming, but is an ambiguous definition due the to the optional capitalization of the first letter. Some programming styles prefer camel case with the first letter capitalised, others not. For clarity, this article calls the two alternatives upper camel case (initial uppercase letter, also known as Pascal case) and lower camel case (initial lowercase letter, also known as dromedary case). Some people and organizations, notably Microsoft, use the term camel case only for lower camel case. Pascal case means only upper camel case. 

I'm not going to "fix" this here because it's a quote, but I've modified the comment in the `Condition` type to say: "upper CamelCase (PascalCase)" for clarity.

Reply:
Huh. I remember those terms back as far as my Uni days. Good enough.